dimagi/commcare-hq
# Delete Case Types and Properties in Data Dictionary V2

Exact same changes as [this PR](https://github.com/dimagi/commcare-hq/pull/33956), but with the commit history cleaned up a bit (no new code).

## Product Description

There is a new button to delete a case property. This button is contextual and will only appear for case properties that are not being used to store actual case data:
![delete-case-prop-button](https://github.com/dimagi/commcare-hq/assets/122617251/15525cde-c8bc-41e6-b6de-9ef911ae8872)

Clicking on the case property delete button will cause a confirmation modal to pop up:
![Screenshot from 2024-02-07 14-55-23](https://github.com/dimagi/commcare-hq/assets/122617251/1155a653-c614-4245-b8b0-dee49542b58b)

After the user has clicked to confirm deletion, the case property will then be marked as deleted. The user will then need to confirm all changes by clicking "Save" at the top-right corner. This follows the same workflow as deprecating a case property.

Similarly, there is a new button to delete a case type. This button is also contextual and only appears if both the case type and all its case properties are not being used by cases:
![delete-case-type-button](https://github.com/dimagi/commcare-hq/assets/122617251/f7874f7b-15fa-40ef-90bd-6d1e846dae5a)

Clicking the case type delete button will cause a confirmation modal to pop up:
![Screenshot from 2024-02-07 14-57-13](https://github.com/dimagi/commcare-hq/assets/122617251/0b8d3988-6c69-47ec-b6c9-d3bc3f165192)

Once the user has confirmed to delete a case type, the page will be refreshed and the deleted case type will be gone from the left-hand side.

If there is an error with deleting a case type, a generic error message will be shown to the user:
![delete-case-type-error](https://github.com/dimagi/commcare-hq/assets/122617251/1fb92a6a-68c3-4637-b5e3-5c152c228746)


## Technical Summary

Link to ticket [here](https://dimagi-dev.atlassian.net/browse/SC-3076).

This adds in the ability for a user to delete case types/properties. It should be noted however, that this only applies to case types/properties that have no associated case data. Once a case starts using a specific case type/property, the type/property can then no longer be deleted in the Data Dictionary. The only way to achieve deleting it would be to first remove the relevant cases.

Note: This PR also contains the following PRs which have been merged into it:
- [Fix Delete Case Type/Prop Errors for Data Dictionary](https://github.com/dimagi/commcare-hq/pull/34073)
- [Change Case Property Delete Button](https://github.com/dimagi/commcare-hq/pull/34085)

The above PRs have been separately reviewed and approved.

## Feature Flag

None.

## Safety Assurance

### Safety story

- Local testing.
- Has gone through QA.
- Automated tests.

### Automated test coverage


Automated tests have been created for the new view and util functions for deleting a case type. Furthermore, the existing unit tests for updating a case property have been updated to test for deleting case properties as well.

### QA Plan


QA has passed. Ticket available [here](https://dimagi-dev.atlassian.net/browse/QA-5997).

**Note:** Due to all the new changes/commits, QA will be doing a second pass on this feature. This will include doing some performance testing to ensure that the function responsible for collecting the list of case types/props that are safe to delete does not incur a big performance hit.

### Rollback instructions




### Labels & Review
